internal/integrations/rawtemplate.go

```diff
@@ -0,0 +1,240 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package integrations // import "github.com/elastic/opentelemetry-collector-components/internal/integrations"
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/pipeline"
+)
+
+// RawTemplate implements the Template interface for raw YAML content.
+// Unused components are removed after resolving it, so the variables they contain are not required.
+type RawTemplate struct {
+	source rawYAMLConfig
+}
+
+// NewRawTemplate creates a new RawTemplate from raw YAML content.
+func NewRawTemplate(raw []byte) (*RawTemplate, error) {
+	var source rawYAMLConfig
+	decoder := yaml.NewDecoder(bytes.NewReader(raw))
+	decoder.KnownFields(true)
+	err := decoder.Decode(&source)
+	if err != nil {
+		return nil, fmt.Errorf("invalid integration template format: %w", err)
+	}
+	if err := source.Validate(); err != nil {
+		return nil, fmt.Errorf("integration template validation failed: %w", err)
+	}
+	return &RawTemplate{source: source}, nil
+}
+
+// Resolve resolves the template using a confmap resolver.
+func (t *RawTemplate) Resolve(ctx context.Context, params map[string]any, pipelines []pipeline.ID) (*Config, error) {
+	selectedPipelines := t.source.Pipelines
+	if len(pipelines) > 0 {
+		var err error
+		selectedPipelines, err = selectComponents(t.source.Pipelines, pipelines)
+		if err != nil {
+			return nil, fmt.Errorf("selecting pipelines: %w", err)
+		}
+	}
+	selectedReceivers, err := selectComponents(t.source.Receivers, listReceivers(selectedPipelines))
+	if err != nil {
+		return nil, fmt.Errorf("selecting receivers: %w", err)
+	}
+	selectedProcessors, err := selectComponents(t.source.Processors, listProcessors(selectedPipelines))
+	if err != nil {
+		return nil, fmt.Errorf("selecting processors: %w", err)
+	}
+
+	factory := newConfmapProviderFactory(rawYAMLConfig{
+		Pipelines:  selectedPipelines,
+		Receivers:  selectedReceivers,
+		Processors: selectedProcessors,
+	})
+	resolver, err := newResolver(factory, params)
+	if err != nil {
+		return nil, err
+	}
+	conf, err := resolver.Resolve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var config Config
+	if err := conf.Unmarshal(&config); err != nil {
+		return nil, fmt.Errorf("failed to generate effective configuration: %w", err)
+	}
+	return &config, config.Validate()
+}
+
+func newResolver(factory confmap.ProviderFactory, variables map[string]any) (*confmap.Resolver, error) {
+	settings := confmap.ResolverSettings{
+		URIs: []string{"config:main"},
+		ProviderFactories: []confmap.ProviderFactory{
+			factory,
+			newVariablesProviderFactory(variables),
+		},
+	}
+	return confmap.NewResolver(settings)
+}
+
+type selectableID interface {
+	comparable
+	fmt.Stringer
+}
+
+func selectComponents[C any, ID selectableID](from map[ID]C, selection []ID) (map[ID]C, error) {
```

Comment:
all callers of `selectComponents()` just check if the returned error is nil. but if the returned `map[ID]C` is nil is never checked and might cause issues.

Reply:
The value returned here is only used in `buildMainYML`, that checks if the map is empty using `len`.
Do you think I should return an initialized empty map instead of `nil`?

Reply:
Changed to don't return `nil`.